### ISSUE TITLE
Model & training question

### ISSUE THREAD

**Issue by aranga81** [carla-simulator/imitation-learning#34]
Hi,
I have the following questions. Really appreciate if anyone can answer..!!
- Firstly I understand the control commands are like one hot encoded inputs for--> High level command, int ( 2 Follow lane, 3 Left, 4 Right, 5 Straight). How are these generated - carla outputs these automatically? or  are they looking at the steer values to output these ?
- speed prediction auxiliary output - really cant get my head around why this is helping longitudinal performance?
- Training: do we run the backprop only on the active branch for each frame ?

**Comment by mvpcom** [comment]
Hey @aranga81 

- High-level commands will be generated by a local planner (A* in this case). 
- Although I don't know why this can be helpful, they mentioned that this speed output practically helped in a better and smooth driving particularly for turns. So, there is not any logical proof available. 
- As far as I understood, the answer is yes. There is a mask variable which is a determiner for that active branch. So, the deep learning framework such as Tensorflow can automatically understand how to backpropagate on that specific branch.

**Comment by aranga81** [comment]
hey @mvpcom 
Thanks for your reply !
So for training, control commands (one-hot encoded) should be inputs?  and 
outputs --> active branch+speed_prediction ?
- Any idea how the loss function is tailored for this ?

**Comment by mvpcom** [comment]
Yes @aranga81, the control commands is an input. If you are going to re-train the model, maybe [this](https://github.com/mvpcom/carlaILTrainer) can help as an initializer code. I will do my best to fix the current issues as soon as possible. You are free to pull any changes. However, for the loss function, you only need to define an additive loss function for all branches weighted by that mask variable. Reading this issues, [1](https://github.com/mvpcom/carlaILTrainer/issues/1), [2](https://github.com/carla-simulator/imitation-learning/issues/1), [3](https://github.com/carla-simulator/imitation-learning/issues/19), and [4](https://github.com/carla-simulator/imitation-learning/issues/18) also will be helpful.

**Comment by aranga81** [comment]
@mvpcom @felipecode for tweaked conditional imitation learning i am generating my own carla datasets - Can you help me with how the high level control command from the simulator can be recorded ?

**Comment by SoumiDas** [comment]
Exactly my questions as above.. how are the high level commands generated ?Are we supposed to use the A* algorithm in this case? If so, what's the input to this?

**Comment by MalekWahidi** [comment]
> @mvpcom @felipecode for tweaked conditional imitation learning i am generating my own carla datasets - Can you help me with how the high level control command from the simulator can be recorded ?

Same question here. Can anyone please help with this?
